### DISCUSSION TITLE
Style Guide Topics

### DISCUSSION THREAD

**Discussion by seangwright** [seangwright/kentico-xperience-13-style-guide#1] [Ideas]
# Style Guide Topics

Add anything here you'd like to see a topic but don't have feel like opening an issue for.

Here are some prompts to get you thinking:
- If you have experience working on Kentico Portal Engine applications, what did you need to unlearn when starting on Xperience with MVC?
- Where do you feel your coding and approaches to configuring Xperience vary the most over time?
- What do you do the same in both simple and complex projects? What do you do differently?
- What are decisions you've made on past projects that have led to problematic outcomes that you've had to undo (or wish you could have)?
- What do you start with on every new Xperience application?
- What would help developers (of all skill levels) when starting on an existing Xperience application that they are new to?
- How do you ensure your applications are flexible to change in the future (ex: switching hosting environments, adding new areas of a site)?

**Comment by KenticoDevTrev** [comment]
Probably the biggest hurdle was getting used to "coding it" vs. "let webparts do it all" mentality.  At first it was hard for me to not feel like i was doing a whole bunch of extra work, but after i discovered some shortcuts and just got used to the flow, it doesn't really take any more time than it did.

That being said, leveraging Visual Studio's default syntax may be a good point to keep coding quick:

1. When doing dependency injection on the Constructor, type in the Interface (ctrl+. if missing a namespace), then hit [space] -> [tab] (creates variable) -> [ctrl+.] -> [enter] to create local variable.
2. Using Visual Studio to create functions that don't exist on your interfaces by typing it out then [ctrl+.] -> [enter] to generate it
3. Creating a good folder structure and possibly project structure right off the bat (feature folders)

and of course, having a good baseline to start with *shameless plug*

**Reply by seangwright** [reply]
I agree that devs are going to benefit a lot from learning to use their tools, if they aren't already proficient with them.

There's a bunch of VS setup that we could recommend to help making working w/ a team, and producing consistent code:
- [Code Cleanup on Save](https://marketplace.visualstudio.com/items?itemName=MadsKristensen.CodeCleanupOnSave)
- C# rules in [editorconfig](https://docs.microsoft.com/en-us/visualstudio/ide/create-portable-custom-editor-options?view=vs-2019)
- nullable reference types
- Creating libraries w/ [SourceLink](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/sourcelink) for debugging assemblies built w/ Release
- Using [User Secrets](https://docs.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-5.0&tabs=windows#secret-manager) for local settings
- Using the right tool for the right job and not performing major client-side development in VS and instead using [VS Code](https://code.visualstudio.com/), which has far better client-side tooling extensions

There's also minor things that can be easily overlooked, like
- Adding a README.md describing the project w/ instructions on how to get it running
- Keeping an [Architecture Decision Record](https://ardalis.com/getting-started-with-architecture-decision-records/) to make sense of project decisions over time

I definitely **don't** want to include any sample code/baseline project in this repo, but I do think it should link out to those.

**Reply by jakekula** [reply]
Hoping this is relevant to the above.

Going from portal to MVC felt like I was coding again. That's not to say that there wasn't heavy customisation involved on some portal builds, but in general and after so many years of repetition it became such a quick and simple drag/drop exercise. Then MVC came along and I was coding things from scratch. So it was challenging but also broke the monotony and added some excitement and purpose. Greater control over HTML also reduced a lot of friction between the back-end and front-end teams.

Having said that, it quickly became apparent that there were some standard repeated elements that always needed to be included, for most builds. This were:
- custom routing handler (pre K13) to allow URLs based on tree nodes
- navigation functionality
- search functionality
- home page and internal page
- a blog/article page with listing page and teaser component
- online form

A generic template was created with these elements included and we called it the Standard Project. So any new site build would basically use this boilerplate and initial dev time was cut down a fair bit.

**Reply by seangwright** [reply]
@jakekula I've heard this many times from devs switching from Portal Engine to MVC.

That's part of the motivation for creating this style guide - lots of devs working with Xperience haven't been doing from-scratch software design recently and now they're being asked to do exactly that.

I know based on my past experience, if I see a pattern somewhere and it 'works' the first time I try it, I'm sticking with it and using it even where it's not a good idea.

I hope this style guide can get in front of Xperience devs before they adopt patterns that might not be the best approach (or at least the guide can inform them of the risks).

Since the community is already creating some boilerplate/template projects, we can link out to those in the guide.
